### PR TITLE
Improvements to initialization

### PR DESCRIPTION
**Pull request by dantleech**
- Intiailizer create workspaces for all configured PHPCR sessions.
- Added purger
- Deprecated sulu:webspaces:init, sulu:phpcr:init
- Replaced by new sulu:document:initialize
- Better-looking intializer output.

```
Usage:
  sulu:document:initialize [options]

Options:
      --purge              Purge the content repository before initialization.
      --force              Do not ask for confiration.
      ...

Help:
 Initialize the PHPCR content repository. This command
 will all registered initializers. 
 
     $ /usr/local/bin/testconsole sulu:document:initialize
     
 WARNING: Initializers SHOULD be idempotent and it SHOULD be safe to run this
          command multiple times - but as we have no control over which initializers are
          registered and what they do this cannot be guaranteed, so use at your own
          discretion on a system that has sensitive data.
```

### CODE REVIEW

File: src/Sulu/Bundle/TestBundle/Testing/SuluTestCase.php [hunk 1]
```diff
@@ -139,8 +139,12 @@ protected function createWebsiteClient()
     }
 
     /**
-     * Initialize / reset the Sulu PHPCR environment
-     * NOTE: This should use initializers when we implement that feature.
+     * Initialize / reset the Sulu PHPCR environment.
+     *
+     * NOTE: We could use the document initializer here rathan manually creating
```
Comment (dantleech) [inline src/Sulu/Bundle/TestBundle/Testing/SuluTestCase.php]: Rather

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@ CHANGELOG for Sulu
 ==================
 
 * dev-develop
+    * FEATURE     #2056 [All]                 Better content repository initialization, deprecated sulu:phpcr:init & sulu:webspaces:init
```
Comment (dantleech) [inline CHANGELOG.md]: Now it is 2076

**Comment by dantleech** [comment]
@danrot if this is OK, let me know and then I will fix the outstanding comments and merge.

File: .travis.yml [hunk 3]
```diff
@@ -19,7 +19,6 @@ before_script:
   - composer self-update
   - composer install
   - ./bin/jackrabbit.sh
-  - php bin/jackrabbit.php
```
Comment (danrot) [inline .travis.yml]: Why is this removed?

**Comment by danrot** [comment]
The `sulu:phpcr:init` and `sulu:webspaces:init` are not doing the same as they did before, do they? If even the behavior changed, I would not keep these commands.

File: src/Sulu/Bundle/ContentBundle/Tests/Functional/Controller/NodeControllerFieldsTest.php [hunk 4]
```diff
@@ -500,8 +470,7 @@ public function testGetTreeActionSingleWebspaceNodes()
 
     public function testGetTreeActionAllWebspaceNodes()
     {
-        $this->initPhpcr();
-        $this->createWebspaceRoot('test_io');
+        $this->createHomeDocument('test_io', ['de']);
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Tests/Functional/Controller/NodeControllerFieldsTest.php]: Where does this method suddenly come from?

File: src/Sulu/Bundle/DocumentManagerBundle/Initializer/Initializer.php [hunk 5]
```diff
@@ -15,25 +15,64 @@
 use Symfony\Component\Console\Output\OutputInterface;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 
+/**
+ * Content repository initializer.
+ *
+ * Can purge and initialize the content repository using any number of
+ * registered *initializers*.
+ *
+ * Intitializers should be designed to be idempotent - they should not do
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Initializer/Initializer.php]: Initializers

File: .travis.yml (hunk 3 above)
Comment (dantleech) [inline .travis.yml]: Its the travis file in the search bundle which was already out of date.

Better to add .travis.yml files if and when we need them.

**Comment by danrot** [comment]
I'd say the commands should do the same as before. If you have fixed these comments I would test it out before merging it.

File: src/Sulu/Bundle/ContentBundle/Tests/Functional/Controller/NodeControllerFieldsTest.php (hunk 4 above)
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Tests/Functional/Controller/NodeControllerFieldsTest.php]: It suddenly comes from SuluTestCase. It creates the webspace root (home document) using the document manager (rather than directly using PHPCR)

**Comment by dantleech** [comment]
The commands are already doing the same before in this PR. My mistake.

File: .travis.yml (hunk 3 above)
Comment (dantleech) [inline .travis.yml]: This file simply created the workspace (manually instantiating the PHPCR connection). This is now handled by the initializer.

File: src/Sulu/Bundle/DocumentManagerBundle/Initializer/Initializer.php [hunk 6]
```diff
@@ -15,25 +15,64 @@
...
+            $output->write('!! Purging workspaces');
+            $this->purger->purge();
+            $output->writeln(' [OK]');
+        }
+
         arsort($this->initializerMap);
 
         foreach (array_keys($this->initializerMap) as $initializerId) {
+            $output->writeln(sprintf('<comment>%s</>', $initializerId));
             $initializer = $this->container->get($initializerId);
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Initializer/Initializer.php]: Wouldn't it be better to collect these services using a CompilerPass, instead of injecting the container here?

File: src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Initialalizer/InitializerTest.php [hunk 7]
```diff
@@ -13,9 +13,13 @@
 
 use Sulu\Bundle\DocumentManagerBundle\Initializer\Initializer;
 use Sulu\Bundle\DocumentManagerBundle\Initializer\InitializerInterface;
+use Sulu\Bundle\DocumentManagerBundle\Initializer\PurgerInterface;
 use Symfony\Component\Console\Output\NullOutput;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 
+/**
+ * TODO: Test purge.
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Initialalizer/InitializerTest.php]: Still relevant?

**Comment by danrot** [comment]
The `sulu:build` command is not working for me: 
image

File: src/Sulu/Bundle/DocumentManagerBundle/Initializer/Initializer.php (hunk 6 above)
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Initializer/Initializer.php]: The idea is to lazily instantiate them. 

The problem is that the command is intantiated every time you launch the
console, so if we were to inject the services then the Initializer would
be instantatiated every time, and then every single initializer class
would also be intnatiated, even though they are not needed.

As it is, only the Initializer service will be instantiated.

**Comment by dantleech** [comment]
Can you see why? That image seems to contradict itself.

**Comment by danrot** [comment]
I can't see why after a quick look. Funny is, that if I execute the exact same command on its own it is working. Does that also happen on your machine?

**Comment by dantleech** [comment]
Was because the options needed to be passed as an array not a string. Updated.

**Comment by danrot** [comment]
@dantleech If you fix https://github.com/sulu-io/sulu/pull/2076#discussion_r54572862 and https://github.com/sulu-io/sulu/pull/2076#discussion_r55512228 I would merge that, after testing again.

**Comment by danrot** [comment]
And of course your own comments about typos :-)

**Comment by dantleech** [comment]
Updated and rebased on develop.
